File: classlink-oauth2/csharp/OneClick.cs
using System;
using System.Collections.Generic;
using System.Collections.Specialized;
using System.IO;
using System.Linq;
using System.Net;
using System.Text;
using Newtonsoft.Json;

namespace ClassLink_OneClick
{
    public class OneClick
    {
        private static readonly string authUrl = "https://launchpad.classlink.com/oauth2/v2/auth";
        private static readonly string exchangeUrl = "https://launchpad.classlink.com/oauth2/v2/token";
        private static readonly string infoUrl = "https://nodeapi.classlink.com/";
        private readonly string clientId;
        private readonly string clientSecret;

        ///<summary>
        /// Create an instance of the ClassLink OneClick helper
        ///</summary>
        ///<param name="client_Id"> Client Identifier string.</param>
        ///<param name="client_Secret"> Unique passphrase to go with the client_Id.</param>
        public OneClick(string client_Id, string client_Secret)
        {
            clientId = client_Id;
            clientSecret = client_Secret;
        }

        /// <summary>
        /// Get the URL where the client can get their access code.
        /// </summary>
        /// <param name="scope">
        /// The scope part of the query string in the returned URL. Cannot be null, will default to <value>"profile"</value> if left out.
        ///     <list type="bullet">
        ///         <listheader>
        ///             Scopes that can be used:
        ///         </listheader>
        ///         <item>
        ///             <term>profile</term>
        ///             <description>Access to user info and identity.</description>
        ///         </item>
        ///         <item>
        ///             <term>oneroster</term>
        ///             <description>Access to oneroster info and classes.</description>
        ///         </item>
        ///         <item>
        ///             <term>full</term>
        ///             <description>Full access to all APIs.</description>
        ///         </item>
        ///     </list>
        /// </param>
        /// <param name="redirectUri">Where the URL will take you upon providing correct login information.  Cannot be null, if left out will default to <value>http://localhost:8080/code</value> </param>
        /// <returns>Returns a plain text URL which can be used to octain an access code.</returns>
        public string getCodeUrl(string scope = "profile", string redirectUri = "http://localhost:8080/code")
        {
            NameValueCollection queryString = System.Web.HttpUtility.ParseQueryString(string.Empty);
            queryString["client_id"] = this.clientId;
            queryString["scope"] = scope;
            queryString["redirect_uri"] = redirectUri;
            queryString["response_type"] = "code";
            return authUrl + "?" + queryString.ToString();
        }

        /// <summary>
        /// Get the user's bearer token which will provide access to the other functions.
        /// </summary>
        /// <param name="code">The client access code obtained by <see cref="getCodeUrl(string,string)"/>.</param>
        /// <returns>Returns a bearer token string</returns>
        public string getToken(string code)
        {
            NameValueCollection queryString = System.Web.HttpUtility.ParseQueryString(string.Empty);
            queryString["client_id"] = clientId;
            queryString["client_secret"] = clientSecret;
            queryString["code"] = code;
            var data = Encoding.ASCII.GetBytes(queryString.ToString());

            HttpWebRequest req = (HttpWebRequest) WebRequest.Create(exchangeUrl);
            req.Method = "POST";
            req.ContentType = "application/x-www-form-urlencoded";
            req.ContentLength = data.Length;

            using(var stream = req.GetRequestStream())
            {
                stream.Write(data, 0, data.Length);
            }

            HttpWebResponse resp = (HttpWebResponse)req.GetResponse();
            string json = new StreamReader(resp.GetResponseStream()).ReadToEnd();

            Dictionary<string, string> resp_vals = JsonConvert.DeserializeObject<Dictionary<string, string>>(json);
            return resp_vals["access_token"];
        }

        /// <summary>
        /// Get information from the requested endpoint about the user(s) or district
        /// </summary>
        /// <param name="bearer">The bearer token returned by <see cref="getToken(string)"/>.</param>
        /// <param name="endpoint">The URL of the server checked for user validation</param>
        /// <param name="extractNodes">
        /// The peice of user info returned.  If it contains multiple items the
        /// returned string will be a json (or json list) containing the items.
        /// If only a single item the returned string will be a single value or
        /// list of values.  If null or left out the entire JSON will be returned.
        /// </param>
        /// <returns>Returns a string containing the user/client info requested, or a JSON containing all of the available user/client info.</returns>
        public string getInfo(string bearer, string endpoint, string[] extractNodes = null)
        {
            HttpWebRequest req = (HttpWebRequest) WebRequest.Create(infoUrl + endpoint);
            req.Method = "GET";
            req.Headers.Set("Authorization", "bearer " + bearer);
            HttpWebResponse resp = (HttpWebResponse)req.GetResponse();
            string json = new StreamReader(resp.GetResponseStream()).ReadToEnd();

            if (extractNodes != null)
            {
                if (extractNodes.Length == 1)
                {
                    if (json[0] == '[')
                    {
                        List<Dictionary<string, string>> resp_vals = JsonConvert.DeserializeObject<List<Dictionary<string, string>>>(json);
                        List<string> rtn = new List<string>();
                        foreach (Dictionary<string, string> d in resp_vals)
                        {
                            rtn.Add(d[extractNodes[0]]);
                        }
                        json = JsonConvert.SerializeObject(rtn/*, Formatting.Indented*/, Formatting.None);
                    }
                    else if (json[0] == '{')
                    {
                        return JsonConvert.DeserializeObject<Dictionary<string, string>>(json)[extractNodes[0]];
                    }
                    else
                    {
                        // invalid response like expired token :(
                        Console.Error.WriteLine(json);
                        return null;
                    }
                }
                else
                {
                    if (json[0] == '[')
                    {
                        List<Dictionary<string, string>> resp_vals = JsonConvert.DeserializeObject<List<Dictionary<string, string>>>(json);
                        List<Dictionary<string, string>> rtn = new List<Dictionary<string, string>>();
                        Dictionary<string, string> tmp = new Dictionary<string, string>();
                        foreach (Dictionary<string, string> d in resp_vals)
                        {
                            tmp.Clear();
                            foreach (KeyValuePair<string, string> kvp in d)
                            {
                                if (extractNodes.Contains(kvp.Key))
                                {
                                    tmp.Add(kvp.Key, kvp.Value);
                                }
                            }
                            rtn.Add(tmp);
                        }

                        json = JsonConvert.SerializeObject(rtn/*, Formatting.Indented*/, Formatting.None);
                    }
                    else if (json[0] == '{')
                    {
                        Dictionary<string, string> resp_val = JsonConvert.DeserializeObject<Dictionary<string, string>>(json);
                        Dictionary<string, string> rtn = new Dictionary<string, string>();
                        foreach (KeyValuePair<string, string> kvp in resp_val)
                        {
                            if (extractNodes.Contains(kvp.Key))
                            {
                                rtn.Add(kvp.Key, kvp.Value);
                            }
                        }

                        json = JsonConvert.SerializeObject(rtn/*, Formatting.Indented*/, Formatting.None);
                    }
                    else
                    {
                        // invalid response like expired token :(
                        Console.Error.WriteLine(json);
                        return null;
                    }
                }
            }
            return json;
        }

        /// <summary>
        /// Get information about the user as a JSON string.
        /// </summary>
        /// <param name="bearer">The bearer token returned by <see cref="getToken(string)"/>.</param>
        /// <returns>Returns a JSON string containing all of the available user info.</returns>
        public string getUserInfo(string bearer) => getInfo(bearer, "v2/my/info");

        /// <summary>
        /// Get the user's district.
        /// </summary>
        /// <param name="bearer">The bearer token returned by <see cref="getToken(string)"/>.</param>
        /// <returns>Returns the district as a string. </returns>
        public string getUserDistrict(string bearer) => getInfo(bearer, "v2/my/district");

        /// <summary>
        /// Get info opn the user's profile(s).
        /// </summary>
        /// <param name="bearer">The bearer token returned by <see cref="getToken(string)"/>.</param>
        /// <returns>Returns a string containing list of JSONs of the profile(s) information.</returns>
        public string getUserProfiles(string bearer) => getInfo(bearer, "v2/my/profiles");

        /// <summary>
        /// If the users is a parent who has registered for Classlink, this will get a list of any linked student accounts.
        /// </summary>
        /// <param name="bearer">The bearer token returned by <see cref="getToken(string)"/>.</param>
        /// <returns>Returns a list JSONs containing all linked information on students.  Returns an empty list if the user is not a parent.</returns>
        public string getUserChildren(string bearer) => getInfo(bearer, "v2/my/students");

        /// <summary>
        /// Get a list of all groups that the user is a part of.
        /// </summary>
        /// <param name="bearer">The bearer token returned by <see cref="getToken(string)"/>.</param>
        /// <returns>Returns a string containing a list of JSONs with the name, groupId, and buildingId of the group.</returns>
        public string getUserGroups(string bearer) => getInfo(bearer, "my/groups");

    }
}

File: classlink-oauth2/java/OneClick.java
/**
 * this package requires you to download and install 2 libraries for it to run
 * Apache http components https://hc.apache.org/downloads.cgi
 * Maven org.json https://mvnrepository.com/artifact/org.json/json
 */

import com.sun.javaws.exceptions.InvalidArgumentException;
import org.apache.http.HttpResponse;
import org.apache.http.client.HttpClient;
import org.apache.http.client.entity.UrlEncodedFormEntity;
import org.apache.http.client.methods.HttpGet;
import org.apache.http.client.methods.HttpPost;
import org.apache.http.client.utils.URLEncodedUtils;
import org.apache.http.impl.client.DefaultHttpClient;
import org.apache.http.NameValuePair;
import org.apache.http.message.BasicNameValuePair;

import java.io.*;
import java.util.ArrayList;
import java.util.List;

import org.json.*;

import javax.xml.ws.http.HTTPException;

public class OneClick {
	private static final String authUrl = "https://launchpad.classlink.com/oauth2/v2/auth";
    private static final String exchangeUrl = "https://launchpad.classlink.com/oauth2/v2/token";
    private static final String infoUrl = "https://nodeapi.classlink.com/";

	private final String clientId;
	private final String clientSecret;

	public final String SCOPE_ONEROSTER = "oneroster";
	public final String SCOPE_FULL = "full";
	public final String SCOPE_PROFILE = "profile";

	/**
	 * ClassLink OneClick helper
	 * @param clientId Client Identifier string.
	 * @param clientSecret Secret to go with the client_Id.
	 */
	public OneClick(String clientId, String clientSecret) {
		this.clientId = clientId;
		this.clientSecret = clientSecret;
	}

	/**
	 * Get the URL for where the client can get their access code
	 * @param scope The scope part of the query string in the returned URL. There are 3 scopes that can be used:
	 *      - profile:  Access to user identity and user specific information info.
	 *      - oneroster: Access to oneroster info and classes.
	 *      - full: Full access to all public APIs.
	 * @param redirectUri Where the URL will take you upon providing correct login information.  Cannot be null, if left out will default to http://localhost:8080/code
	 * @return Returns a URL that the user can curl to or that can be put URL bar.
	 */
	public String getCodeUrl(String scope, String redirectUri) {
		List<NameValuePair> query  = new ArrayList<NameValuePair>();
		query.add(new BasicNameValuePair("client_id", this.clientId));
		query.add(new BasicNameValuePair("scope", scope));
		query.add(new BasicNameValuePair("redirect_uri", redirectUri));
		query.add(new BasicNameValuePair("response_type", "code"));

		String queryString = URLEncodedUtils.format(query, "UTF-8");

		return this.authUrl + "?" + queryString;
	}
	public String getCodeUrl() { return this.getCodeUrl("profile", "https://localhost:8080"); }
	public String getCodeUrl(String scope) { return this.getCodeUrl(scope, "https://localhost:8080"); }

	public String getToken(String code) throws UnsupportedEncodingException, IOException, InvalidArgumentException{
		List<NameValuePair> body  = new ArrayList<NameValuePair>();
		body.add(new BasicNameValuePair("client_id", this.clientId));
		body.add(new BasicNameValuePair("client_secret", this.clientSecret));
		body.add(new BasicNameValuePair("code", code));

		HttpClient client = new DefaultHttpClient();
		HttpPost post = new HttpPost(this.exchangeUrl);

		post.setEntity(new UrlEncodedFormEntity(body));
		HttpResponse response = client.execute(post);
		BufferedReader rd = new BufferedReader(new InputStreamReader(response.getEntity().getContent()));

		StringBuffer result = new StringBuffer();
		String line = "";
		while ((line = rd.readLine()) != null) {
			result.append(line);
		}

		int statusCode = response.getStatusLine().getStatusCode();

		if (statusCode != 200) {
			System.err.println(result.toString());
			throw new HTTPException(statusCode);
		}
		else {
			JSONObject json = new JSONObject(result.toString());
			return json.getString("access_token");
		}
	}

	/**
	 * Get information from the requested endpoint about the user(s) or district
	 * @param bearer The bearer token returned by getToken(String code).
	 * @param endPoint The URI of the server checked for user validation.
	 * @param extractNode The piece of user info returned.  If null or left out the entire JSON will be returned.
	 * @return If extractNode is specified it will return the information in that node as a string.  If extractNode is not specified it will return the JSON encoded response as a string.
	 */
	public String getInfo(String bearer, String endPoint, String extractNode) throws IOException{
		String url = this.infoUrl + endPoint;

		HttpClient client = new DefaultHttpClient();
		HttpGet request = new HttpGet(url);
		request.addHeader("Authorization", "bearer " + bearer);
//		System.out.println(bearer);
//		if(bearer.substring(0, 4) == "bearer".substring(0 ,4)){
//			request.addHeader("Authorization", bearer);
//		}
//		else {
//			request.addHeader("gwstoken", bearer);
//		}

		HttpResponse response = client.execute(request);

		BufferedReader rd = new BufferedReader(new InputStreamReader(response.getEntity().getContent()));
		StringBuffer result = new StringBuffer();
		String line = "";
		while ((line = rd.readLine()) != null) {
			result.append(line);
		}

		int statusCode = response.getStatusLine().getStatusCode();

		if(statusCode != 200) {
			System.err.println(result.toString());
			throw new HTTPException(statusCode);
		}
		else {
			if(extractNode != null) {
				JSONObject json = new JSONObject(result.toString());
				return json.get(extractNode).toString();	//get and not getJSONObject or getJSONArray because it could be both and type errors are a thing
			}
			return result.toString();
		}
	}
	public String getInfo(String bearer, String endPoint) throws IOException {return this.getInfo(bearer, endPoint, null);}

	/**
	 * Get information about the user as a JSON.
	 * @param bearer The bearer token returned by getToken(String code).
	 * @return Returns a JSON containing all of the available user info.
	 * @throws IOException
	 */
	public String getUserInfo(String bearer) throws IOException { return this.getInfo(bearer, "v2/my/info"); }

	/**
	 * Get the district associated with the user.
	 * @param bearer The bearer token returned by getToken(String code).
	 * @return Returns the district as a string.
	 * @throws IOException
	 */
	public String getUserDistrict(String bearer) throws IOException { return this.getInfo(bearer, "v2/my/district");}

	/**
	 * Get info on the user's profile(s).
	 * @param bearer The bearer token returned by getToken(String code).
	 * @return Returns a list of JSONs containing the profile(s) information as a string.
	 * @throws IOException
	 */
	public String getUserProfiles(String bearer) throws IOException { return this.getInfo(bearer, "v2/my/profiles"); }

	/**
	 * If the user is a parent who has registered for Classlink this will get a list of any linked student accounts aka their children.
	 * @param bearer The bearer token returned by getToken(String code).
	 * @return Returns a list JSONs containing all linked information on students. Returns an empty list if the user is not a parent.
	 * @throws IOException
	 */
	public String getUserChildren(String bearer) throws IOException { return this.getInfo(bearer, "v2/my/students"); }

	/**
	 * Get a list of all groups that the user is a part of.
	 * @param bearer The bearer token returned by getToken(String code).
	 * @return Returns a list JSONs containing the name, groupId, and buildingId of the group.
	 * @throws IOException
	 */
	public String getUserGroups(String bearer) throws IOException { return this.getInfo(bearer, "my/groups"); }

}
